File: backend/services/stigmer-server/pkg/domain/workflow/registry/model_registry_store.go
package registry

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"sort"
	"sync"
	"time"

	"github.com/rs/zerolog/log"
)

// Where the store refreshes from, and how it is controlled. The cloud
// endpoint is public and unauthenticated (raw provider list prices only —
// DD-004); self-hosted operators can repoint or disable it entirely.
const (
	// EnvModelRegistryUpstream overrides the cloud origin the store
	// refreshes from (default: the hosted Stigmer API).
	EnvModelRegistryUpstream = "STIGMER_MODEL_REGISTRY_UPSTREAM"

	// EnvModelRegistryRefresh set to "off" disables the background
	// refresh entirely — the bundled registry serves forever. Air-gapped
	// installs work either way (a failed refresh quietly keeps the
	// bundle); "off" additionally guarantees zero outbound calls.
	EnvModelRegistryRefresh = "STIGMER_MODEL_REGISTRY_REFRESH"

	defaultModelRegistryUpstream = "https://api.stigmer.ai"
	publicModelRegistryPath      = "/api/v1/public/model-registry"

	modelRegistryRefreshInterval = time.Hour
	modelRegistryFetchTimeout    = 30 * time.Second

	// A registry document is a few hundred KB; anything near this bound
	// is not a registry.
	maxModelRegistryBytes = 8 << 20
)

// ModelRegistryStore holds the current model-registry document and the
// valid-model indexes derived from it.
//
// Both registry consumers in this server — the /v1/proxy/model-registry
// HTTP handler and workflow model validation — read this one store, so
// the document the pickers see and the set validation accepts can never
// drift (DD-004: a model added in the cloud console must not pass the
// picker and then fail workflow validation).
//
// The bundled (embedded) registry loads at construction: the store works
// fully offline and no startup path touches the network. StartRefresh
// upgrades it — a background loop pulls the public cloud endpoint and,
// when a response passes the sanity gate, swaps the document and indexes
// atomically under the write lock.
type ModelRegistryStore struct {
	mu                    sync.RWMutex
	document              []byte
	modelsByHarness       map[string]map[string]bool
	sortedModelsByHarness map[string][]string

	// Pricing-variant capability index (stigmer/stigmer#357): variant key
	// ("fast") → harness → model references that price it. Presence of a
	// priced variant IS the selectability capability — a tier with no price
	// would trip billing's undercharge guard, so selection and billability
	// are coupled by construction. The harness dimension exists for the
	// workflow validators (which know the task's harness); the execution
	// create step stays deliberately harness-free (it never resolves the
	// session) and unions across harnesses.
	modelsByVariant map[string]map[string]map[string]bool
	// variant key → sorted canonical ids across all harnesses (create-step
	// refusal messages).
	sortedModelsByVariant map[string][]string
	// variant key → harness → sorted canonical ids (workflow-validation
	// refusal messages).
	sortedModelsByVariantHarness map[string]map[string][]string

	// Log-noise control: the first refresh failure warns, subsequent
	// consecutive failures log at debug until a refresh succeeds again.
	failureLogged bool
}

// FastVariantKey is the registry pricing-variant key backing
// SERVICE_TIER_FAST (stigmer/stigmer#357) — the one vocabulary shared by
// every consumer (execution create validation, workflow validation), defined
// beside the index it queries so the selectable key and the priced key can
// never drift.
const FastVariantKey = "fast"

// The subset of a registry entry the store indexes. Both canonical ids
// and provider api ids are accepted as valid references because the
// runner resolves canonical ids via the registry and passes
// unknown-but-registered api ids to the provider verbatim
// (stigmer/stigmer#240). PricingVariants values are deliberately not
// modeled — only the key set matters for capability.
type modelRegistryEntry struct {
	ID              string                     `json:"id"`
	ApiModelID      string                     `json:"apiModelId"`
	Harness         string                     `json:"harness"`
	PricingVariants map[string]json.RawMessage `json:"pricingVariants"`
}

type modelRegistryData struct {
	Models []modelRegistryEntry `json:"models"`
}

var (
	storeOnce sync.Once
	store     *ModelRegistryStore
)

// Store returns the process-wide model registry store, initialized from
// the embedded bundle on first use. Fatal if the build-time bundle is
// missing — a server without any registry cannot resolve models at all.
func Store() *ModelRegistryStore {
	storeOnce.Do(func() {
		data, err := registryFS.ReadFile("data/model-registry.json")
		if err != nil {
			log.Fatal().Err(err).Msg("Failed to load the bundled model-registry.json — " +
				"the embed is created at build time (refresh it with 'make sync-model-registry')")
		}
		s := &ModelRegistryStore{}
		if err := s.applyDocument(data); err != nil {
			log.Fatal().Err(err).Msg("Bundled model-registry.json is not a valid registry document")
		}
		log.Info().Int("bytes", len(data)).Msg("Loaded bundled model-registry.json")
		store = s
	})
	return store
}

// Document returns the current registry document bytes (bundle or the
// last good upstream response). Callers must not mutate the slice.
func (s *ModelRegistryStore) Document() []byte {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.document
}

// IsValidModel reports whether a model reference (canonical id or
// provider api id) is executable on the given harness.
func (s *ModelRegistryStore) IsValidModel(harness, model string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.modelsByHarness[harness][model]
}

// HasHarness reports whether the registry knows any models for a harness.
func (s *ModelRegistryStore) HasHarness(harness string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.modelsByHarness[harness]) > 0
}

// HasAnyModels reports whether the registry loaded at all — validation
// degrades to a no-op rather than rejecting everything when it did not.
func (s *ModelRegistryStore) HasAnyModels() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.modelsByHarness) > 0
}

// IsValidModelOnAnyHarness reports whether a model reference (canonical id
// or provider api id) is executable on AT LEAST ONE harness. The
// existence check for surfaces with no serving harness in this edition
// (agent channels): a pin no section knows is certainly a typo, while a
// pin valid anywhere may be right where the spec actually serves.
func (s *ModelRegistryStore) IsValidModelOnAnyHarness(model string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	for _, models := range s.modelsByHarness {
		if models[model] {
			return true
		}
	}
	return false
}

// CanonicalModelsAcrossHarnesses returns the sorted, deduplicated
// canonical model ids across every harness section — the did-you-mean
// candidate pool for the any-harness existence check. Computed per call
// (refusal-path only); callers may keep the slice.
func (s *ModelRegistryStore) CanonicalModelsAcrossHarnesses() []string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	seen := make(map[string]bool)
	var merged []string
	for _, models := range s.sortedModelsByHarness {
		for _, name := range models {
			if !seen[name] {
				seen[name] = true
				merged = append(merged, name)
			}
		}
	}
	sort.Strings(merged)
	return merged
}

// CanonicalModels returns the sorted canonical model ids for a harness
// (for deterministic did-you-mean suggestions — canonical ids are the
// documented form, so suggestions never surface provider api ids).
// Callers must not mutate the slice.
func (s *ModelRegistryStore) CanonicalModels(harness string) []string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.sortedModelsByHarness[harness]
}

// HasPricingVariant reports whether a model reference (canonical id or
// provider api id) prices the given variant key (e.g. "fast") under ANY
// harness — the registry-backed capability check for
// ExecutionConfig.service_tier at execution create (stigmer/stigmer#357),
// which is deliberately harness-free (it never resolves the session).
func (s *ModelRegistryStore) HasPricingVariant(model, variant string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	for _, refs := range s.modelsByVariant[variant] {
		if refs[model] {
			return true
		}
	}
	return false
}

// HasPricingVariantForHarness reports whether a model reference prices the
// given variant key under the given harness. The workflow validators use
// this form — the task config names its harness, so a fast variant priced
// only under another harness must not validate (it would execute as a
// silent no-op).
func (s *ModelRegistryStore) HasPricingVariantForHarness(harness, model, variant string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.modelsByVariant[variant][harness][model]
}

// CanonicalModelsWithVariant returns the sorted canonical model ids that
// price the given variant key under any harness, for actionable refusal
// messages ("fast is available on: ..."). Callers must not mutate the
// slice.
func (s *ModelRegistryStore) CanonicalModelsWithVariant(variant string) []string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.sortedModelsByVariant[variant]
}

// CanonicalModelsWithVariantForHarness returns the sorted canonical model
// ids that price the given variant key under the given harness. Callers
// must not mutate the slice.
func (s *ModelRegistryStore) CanonicalModelsWithVariantForHarness(harness, variant string) []string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.sortedModelsByVariantHarness[variant][harness]
}

// applyDocument sanity-gates and installs a registry document: it must
// parse, and it must index at least one model. A malformed or empty
// upstream response never replaces a working registry.
func (s *ModelRegistryStore) applyDocument(data []byte) error {
	var parsed modelRegistryData
	if err := json.Unmarshal(data, &parsed); err != nil {
		return fmt.Errorf("parse model registry document: %w", err)
	}

	byHarness := make(map[string]map[string]bool)
	canonicalByHarness := make(map[string][]string)
	byVariant := make(map[string]map[string]map[string]bool)
	// Set-backed: the same canonical id may price a variant under more than
	// one harness, and the union list must not repeat it.
	canonicalByVariantSet := make(map[string]map[string]bool)
	canonicalByVariantHarness := make(map[string]map[string][]string)
	for _, m := range parsed.Models {
		// $comment section dividers carry no id/harness.
		if m.ID == "" || m.Harness == "" {
			continue
		}
		if byHarness[m.Harness] == nil {
			byHarness[m.Harness] = make(map[string]bool)
		}
		byHarness[m.Harness][m.ID] = true
		canonicalByHarness[m.Harness] = append(canonicalByHarness[m.Harness], m.ID)
		if m.ApiModelID != "" {
			byHarness[m.Harness][m.ApiModelID] = true
		}
		for variant := range m.PricingVariants {
			if byVariant[variant] == nil {
				byVariant[variant] = make(map[string]map[string]bool)
			}
			if byVariant[variant][m.Harness] == nil {
				byVariant[variant][m.Harness] = make(map[string]bool)
			}
			byVariant[variant][m.Harness][m.ID] = true
			if m.ApiModelID != "" {
				byVariant[variant][m.Harness][m.ApiModelID] = true
			}
			if canonicalByVariantSet[variant] == nil {
				canonicalByVariantSet[variant] = make(map[string]bool)
			}
			canonicalByVariantSet[variant][m.ID] = true
			if canonicalByVariantHarness[variant] == nil {
				canonicalByVariantHarness[variant] = make(map[string][]string)
			}
			canonicalByVariantHarness[variant][m.Harness] =
				append(canonicalByVariantHarness[variant][m.Harness], m.ID)
		}
	}
	if len(byHarness) == 0 {
		return fmt.Errorf("registry document contains no model entries")
	}
	for _, ids := range canonicalByHarness {
		sort.Strings(ids)
	}
	canonicalByVariant := make(map[string][]string, len(canonicalByVariantSet))
	for variant, idSet := range canonicalByVariantSet {
		ids := make([]string, 0, len(idSet))
		for id := range idSet {
			ids = append(ids, id)
		}
		sort.Strings(ids)
		canonicalByVariant[variant] = ids
	}
	for _, byH := range canonicalByVariantHarness {
		for _, ids := range byH {
			sort.Strings(ids)
		}
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	s.document = data
	s.modelsByHarness = byHarness
	s.sortedModelsByHarness = canonicalByHarness
	s.modelsByVariant = byVariant
	s.sortedModelsByVariant = canonicalByVariant
	s.sortedModelsByVariantHarness = canonicalByVariantHarness
	return nil
}

// StartModelRegistryRefresh begins the background refresh of the
// process-wide store from the public cloud endpoint. Called once by the
// server after boot; nothing else ever initiates network activity from
// this package, so binaries and tests that merely import it stay
// network-silent.
func StartModelRegistryRefresh(ctx context.Context) {
	Store().startRefresh(ctx, os.Getenv(EnvModelRegistryUpstream),
		os.Getenv(EnvModelRegistryRefresh))
}

func (s *ModelRegistryStore) startRefresh(ctx context.Context, upstream, refreshSetting string) {
	if refreshSetting == "off" {
		log.Info().Msg("Model registry refresh disabled (" + EnvModelRegistryRefresh + "=off) — " +
			"serving the bundled registry")
		return
	}
	if upstream == "" {
		upstream = defaultModelRegistryUpstream
	}
	url := upstream + publicModelRegistryPath

	go func() {
		s.refreshOnce(ctx, url)
		ticker := time.NewTicker(modelRegistryRefreshInterval)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				s.refreshOnce(ctx, url)
			}
		}
	}()
}

// refreshOnce fetches the upstream registry and applies it when sane. A
// failure of any kind keeps the current document: the first consecutive
// failure warns (self-hosted installs are often offline — that is a
// supported mode, not an error loop), later ones log at debug.
func (s *ModelRegistryStore) refreshOnce(ctx context.Context, url string) {
	data, err := fetchModelRegistry(ctx, url)
	if err == nil {
		err = s.applyDocument(data)
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	if err != nil {
		if s.failureLogged {
			log.Debug().Err(err).Str("url", url).Msg("Model registry refresh failed; keeping the current registry")
		} else {
			log.Warn().Err(err).Str("url", url).Msg("Model registry refresh failed; keeping the current registry " +
				"(offline self-hosting is fine — set " + EnvModelRegistryRefresh + "=off to silence this)")
			s.failureLogged = true
		}
		return
	}
	if s.failureLogged {
		s.failureLogged = false
	}
	log.Info().Int("bytes", len(data)).Str("url", url).Msg("Model registry refreshed from upstream")
}

func fetchModelRegistry(ctx context.Context, url string) ([]byte, error) {
	fetchCtx, cancel := context.WithTimeout(ctx, modelRegistryFetchTimeout)
	defer cancel()

	req, err := http.NewRequestWithContext(fetchCtx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("upstream returned %d", resp.StatusCode)
	}
	data, err := io.ReadAll(io.LimitReader(resp.Body, maxModelRegistryBytes))
	if err != nil {
		return nil, err
	}
	return data, nil
}

File: backend/services/stigmer-server/src/domain/workflow/registry/model-registry-store.ts
/**
 * Model-registry document store — ports the Go server's ModelRegistryStore
 * (pkg/domain/workflow/registry/model_registry_store.go). Domain-owned, as
 * in Go: the workflow validators and the transport's /v1/proxy/model-registry
 * lane read this ONE store, so the document the pickers see and the set
 * validation accepts can never drift (DD-004). Sub-project #3 first landed
 * the serving/refresh half under src/transport/; the workflow-family port
 * moved it home and added the catalog indexes (sub-project DD-A).
 *
 * The store always holds a complete, valid document: it starts from the
 * bundled build-time snapshot (offline-install contract — a server with no
 * outbound network still answers a full registry) and upgrades in place
 * from the hosted API. Every transition is guarded by the same sanity gate
 * Go applies (applyDocument, :299-306): the candidate must parse and index
 * at least one model entry, otherwise it is rejected and the current
 * document stays — a bad upstream can never blank the console's model
 * pickers or turn model validation into reject-everything.
 *
 * Failure logging matches Go (:454-476): the FIRST consecutive refresh
 * failure logs at warn, repeats at debug until a success resets the flag —
 * an offline laptop must not fill its log with hourly warnings.
 *
 * Since O5 (20260827.02) this class is the OSS implementation of the
 * ModelCatalogProvider seam (DD-008) — consumers hold the interface; the
 * refresh lifecycle below stays composition-owned, outside the contract.
 * The document INTERPRETATION (sanity gate, indexes, the query methods)
 * lives in document-catalog.ts since the C1 seam extraction (20260827.04):
 * this class owns only the bundled/upstream lifecycle and delegates every
 * read to the current DocumentModelCatalog, swapped atomically per
 * accepted document.
 */
import type { Logger } from "../../../boot/logger.js";
import { DocumentModelCatalog } from "./document-catalog.js";
import type { ModelCatalogProvider } from "./model-catalog-provider.js";

/** Upstream refresh cadence (Go modelRegistryRefreshInterval, :34). */
export const MODEL_REGISTRY_REFRESH_INTERVAL_MS = 60 * 60 * 1000;

/** Per-fetch budget (Go fetch timeout, :35-39). */
export const MODEL_REGISTRY_FETCH_TIMEOUT_MS = 30_000;

/**
 * Read cap on the upstream body (Go's 8<<20 limited reader). Go truncates
 * the stream and lets the parse fail; here the size check rejects before
 * parsing — same outcome (oversized document rejected, current kept).
 */
export const MODEL_REGISTRY_MAX_BYTES = 8 * 1024 * 1024;

/** Hosted API path appended to the upstream origin (Go :20-32). */
export const MODEL_REGISTRY_UPSTREAM_PATH = "/api/v1/public/model-registry";

/**
 * The registry pricing-variant key backing SERVICE_TIER_FAST (oss#357) —
 * the one vocabulary shared by every consumer, defined beside the index it
 * queries so the selectable key and the priced key can never drift.
 */
export const FAST_VARIANT_KEY = "fast";

/**
 * The registry capability key backing THINKING_MODE_ENABLED (oss#772) —
 * defined beside the index it queries so the selectable key and the
 * declared key can never drift.
 */
export const THINKING_CAPABILITY_KEY = "thinking";

export interface ModelRegistryStoreOptions {
  /** Bundled build-time snapshot; MUST pass the sanity gate at boot. */
  bundledDocument: string;
  upstreamOrigin: string;
  refreshEnabled: boolean;
  logger: Logger;
  /** Test seam for the upstream fetch. */
  fetchImpl?: typeof fetch;
}

export class ModelRegistryStore implements ModelCatalogProvider {
  private currentCatalog: DocumentModelCatalog;
  private refreshTimer: NodeJS.Timeout | undefined;
  private failureLogged = false;
  private readonly options: ModelRegistryStoreOptions;

  constructor(options: ModelRegistryStoreOptions) {
    // A broken bundled snapshot is a build defect, not a runtime condition:
    // fail loud at construction rather than serving an empty registry
    // (Go log.Fatal in Store()).
    const catalog = DocumentModelCatalog.tryBuild(options.bundledDocument);
    if (catalog === undefined) {
      throw new Error(
        "bundled model-registry snapshot is invalid or indexes no models",
      );
    }
    this.currentCatalog = catalog;
    this.options = options;
  }

  /** The served bytes; always complete and valid by construction. */
  document(): string {
    return this.currentCatalog.document();
  }

  /**
   * Whether a model reference (canonical id or provider api id) is
   * executable on the given harness.
   */
  isValidModel(harness: string, model: string): boolean {
    return this.currentCatalog.isValidModel(harness, model);
  }

  /** Whether the registry knows any models for a harness. */
  hasHarness(harness: string): boolean {
    return this.currentCatalog.hasHarness(harness);
  }

  /**
   * Whether the registry loaded at all — validation degrades to a no-op
   * rather than rejecting everything when it did not.
   */
  hasAnyModels(): boolean {
    return this.currentCatalog.hasAnyModels();
  }

  /**
   * Whether a model reference is executable on AT LEAST ONE harness. The
   * existence check for surfaces with no serving harness in this edition
   * (agent channels): a pin no section knows is certainly a typo, while a
   * pin valid anywhere may be right where the spec actually serves.
   */
  isValidModelOnAnyHarness(model: string): boolean {
    return this.currentCatalog.isValidModelOnAnyHarness(model);
  }

  /**
   * The sorted, deduplicated canonical model ids across every harness
   * section — the did-you-mean candidate pool for the any-harness
   * existence check. Computed per call (refusal-path only).
   */
  canonicalModelsAcrossHarnesses(): string[] {
    return this.currentCatalog.canonicalModelsAcrossHarnesses();
  }

  /**
   * The sorted canonical model ids for a harness (deterministic
   * did-you-mean suggestions — canonical ids are the documented form, so
   * suggestions never surface provider api ids). Callers must not mutate.
   */
  canonicalModels(harness: string): string[] {
    return this.currentCatalog.canonicalModels(harness);
  }

  /**
   * Whether a model reference prices the given variant key under ANY
   * harness — the registry-backed capability check for
   * ExecutionConfig.service_tier at execution create (oss#357), which is
   * deliberately harness-free (it never resolves the session).
   */
  hasPricingVariant(model: string, variant: string): boolean {
    return this.currentCatalog.hasPricingVariant(model, variant);
  }

  /**
   * Whether a model reference prices the given variant key under the given
   * harness. The workflow validators use this form — the task config names
   * its harness, so a fast variant priced only under another harness must
   * not validate (it would execute as a silent no-op).
   */
  hasPricingVariantForHarness(
    harness: string,
    model: string,
    variant: string,
  ): boolean {
    return this.currentCatalog.hasPricingVariantForHarness(
      harness,
      model,
      variant,
    );
  }

  /**
   * The sorted canonical model ids that price the given variant key under
   * any harness, for actionable refusal messages. Callers must not mutate.
   */
  canonicalModelsWithVariant(variant: string): string[] {
    return this.currentCatalog.canonicalModelsWithVariant(variant);
  }

  /**
   * The sorted canonical model ids that price the given variant key under
   * the given harness. Callers must not mutate.
   */
  canonicalModelsWithVariantForHarness(
    harness: string,
    variant: string,
  ): string[] {
    return this.currentCatalog.canonicalModelsWithVariantForHarness(
      harness,
      variant,
    );
  }

  /**
   * Whether a model reference declares the given capability key (e.g.
   * "thinking") true under the given harness. Capability flags are
   * harness-scoped facts (they describe what works through that serving
   * path), so there is no any-harness form: THINKING_MODE_ENABLED validates
   * against the cursor harness specifically — the only harness with a
   * thinking translation in v1 (oss#772) — and native entries declaring the
   * same capability stay unselectable until a native wire mapping exists.
   */
  hasCapabilityForHarness(
    harness: string,
    model: string,
    capability: string,
  ): boolean {
    return this.currentCatalog.hasCapabilityForHarness(
      harness,
      model,
      capability,
    );
  }

  /**
   * The sorted canonical model ids that declare the given capability key
   * under the given harness, for actionable refusal messages.
   */
  canonicalModelsWithCapabilityForHarness(
    harness: string,
    capability: string,
  ): string[] {
    return this.currentCatalog.canonicalModelsWithCapabilityForHarness(
      harness,
      capability,
    );
  }

  /** Immediate refresh, then hourly — Go's goroutine shape (:435-447). */
  startRefresh(): void {
    if (!this.options.refreshEnabled || this.refreshTimer !== undefined) {
      return;
    }
    void this.refreshOnce();
    this.refreshTimer = setInterval(() => {
      void this.refreshOnce();
    }, MODEL_REGISTRY_REFRESH_INTERVAL_MS);
    this.refreshTimer.unref();
  }

  stopRefresh(): void {
    if (this.refreshTimer !== undefined) {
      clearInterval(this.refreshTimer);
      this.refreshTimer = undefined;
    }
  }

  /** Exposed for tests; production goes through startRefresh. */
  async refreshOnce(): Promise<void> {
    const url = this.options.upstreamOrigin + MODEL_REGISTRY_UPSTREAM_PATH;
    const fetchImpl = this.options.fetchImpl ?? fetch;
    try {
      const response = await fetchImpl(url, {
        signal: AbortSignal.timeout(MODEL_REGISTRY_FETCH_TIMEOUT_MS),
      });
      if (!response.ok) {
        throw new Error(`upstream answered ${response.status}`);
      }
      const body = await response.text();
      if (Buffer.byteLength(body) > MODEL_REGISTRY_MAX_BYTES) {
        throw new Error("upstream document exceeds the size cap");
      }
      const catalog = DocumentModelCatalog.tryBuild(body);
      if (catalog === undefined) {
        throw new Error(
          "upstream document failed the sanity gate (no indexable models)",
        );
      }
      // Document and indexes swap together — a reader never sees a
      // document whose indexes describe a different registry.
      this.currentCatalog = catalog;
      this.failureLogged = false;
    } catch (error) {
      const fields = {
        url,
        error: error instanceof Error ? error.message : String(error),
      };
      if (this.failureLogged) {
        this.options.logger.debug(
          "model-registry refresh failed; serving current document",
          fields,
        );
      } else {
        this.options.logger.warn(
          "model-registry refresh failed; serving current document",
          fields,
        );
        this.failureLogged = true;
      }
    }
  }
}
